src/gui/guihelper.cpp

```diff
@@ -0,0 +1,204 @@
+    pVis->AddCamera(chrono::ChVector<>(8, -25, 15), chrono::ChVector<>(0, 0, 0));
+    pVis->AddTypicalLights();
+}
+
+void GUIImpl::SetCamera(double x, double y, double z, double dirx, double diry, double dirz) {
+    pVis->AddCamera({x, y, z}, {dirx, diry, dirz});
+}
+
+bool GUIImpl::IsRunning(double timestep) {
+    if (pVis->Run() == false) return false;
+
+    pVis->BeginScene();
+    pVis->Render();
+
+    // Add grid to materialize horizontal plane
+    tools::drawGrid(pVis.get(), 1, 1, 30, 30,
+                    chrono::ChCoordsys<>(chrono::ChVector<>(0, 0.0, 0), chrono::Q_from_AngZ(chrono::CH_C_PI_2)),
+                    chrono::ChColor(.1f, .1f, .1f), true);
+
+    pVis->EndScene();
+    return true;
+}
+
+hydroc::gui::GUIImpl::MyActionReceiver::MyActionReceiver(bool& buttonPressed) : pressed(buttonPressed) {}
+
+/// @brief Initialize Action with System
+/// @param vsys
+void hydroc::gui::GUIImpl::MyActionReceiver::Init(chrono::irrlicht::ChVisualSystemIrrlicht* vsys) {
+    // store pointer application
+    vis = vsys;
+
+    // ..add a GUI button to control pause/play
+    pauseButton = vis->GetGUIEnvironment()->addButton(rect<s32>(510, 20, 650, 35));
+    buttonText  = vis->GetGUIEnvironment()->addStaticText(L"Paused", rect<s32>(560, 20, 600, 35), false);
+}
+
+bool hydroc::gui::GUIImpl::MyActionReceiver::OnEvent(const irr::SEvent& event) {
+    // check if user clicked button
+    if (event.EventType == EEVENT_TYPE::EET_GUI_EVENT) {
+        switch (event.GUIEvent.EventType) {
+            case EGUI_EVENT_TYPE::EGET_BUTTON_CLICKED:
+                pressed = !pressed;
+                if (pressed) {
+                    buttonText->setText(L"Playing");
+                } else {
+                    buttonText->setText(L"Paused");
+                }
+                return pressed;
+                break;
+            default:
+                break;
+        }
+    }
+    return false;
+}
+
+#else  // HYDROCHRONO_HAVE_IRRLICHT
+
+GUIImpl::GUIImpl() {}
+
+void GUIImpl::Init(UI& ui, chrono::ChSystem* system, const char* title) {
+    std::cout << "Warning: GUI deactivated. Compilation without Irrlicht library" << std::endl;
+}
+
+void GUIImpl::SetCamera(double x, double y, double z, double dirx, double diry, double dirz) {}
+
+bool GUIImpl::IsRunning(double timestep) {
+    return true;
+}
+
+#endif  // HYDROCHRONO_HAVE_IRRLICHT
+
+//
+
+GUI::GUI() : pImpl(std::make_shared<hydroc::gui::GUIImpl>()) {
+    simulationStarted = false;  // Simulation is Paused
+}
+
+void GUI::Init(chrono::ChSystem* system, const char* title) {
+    UI::Init(system, title);
+    pImpl->Init(*this, system, title);
+}
+
+void GUI::SetCamera(double x, double y, double z, double dirx, double diry, double dirz) {
+    pImpl->SetCamera(x, y, z, dirx, diry, dirz);
+}
+
+bool GUI::IsRunning(double timestep) {
+    if (pImpl->IsRunning(timestep) == false) return false;
+
+    // If still running call the base class
+    return UI::IsRunning(timestep);
+}
```

Comment:
I'm confused on why GUIImpl is separate from GUI if GUI's member functions only ever call the GUIImpl functions.